File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int[] result = new int[nums1.length];

        for(int i=0; i<nums1.length; i++) {
            int greaterIdx = -1, j = nums2.length - 1;
            
            while(j >= 0 && nums2[j] != nums1[i]) {
                if(nums2[j] > nums1[i]){
                    greaterIdx = nums2[j];
                }
                
                j--;
            }
            
            result[i] = greaterIdx;
        }

        return result;
    }
}

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        ans = {}
        n = len(nums2)
        st = []
        for i in range(n-1, -1, -1):
            while st and st[-1] <= nums2[i]:
                st.pop()
            if st:
                ans[nums2[i]] = st[-1]
            st.append(nums2[i])
            
        return [ans.get(num, -1) for num in nums1]